src/tracker.rs

```diff
@@ -164,11 +136,12 @@ enum ThresholdAction {
     /// Emit an event. The count of unique tag sets, per metric, is provided as the tags
     Event { title: String, text: String },
 
-    /// Take some custom action. The function will be passed the metric name and tags
+    /// Take some custom action. The function will be passed the metric name and tags, and
+    /// a BTreeMap<String, Vec<BTreeSet<String>>> of all the unique tag sets seen for each metric.
```

Comment:
This is the very last thing: could you move the sentence _"The function will be passed [...] all the unique tag sets seen for each metric."_ to `with_custom_action`?
It's very useful to know for users, but since this is not public API no one (except devs) will see it.

Small note aside: I fear that exposing the `BTreeMap<String, Vec<BTreeSet<String>>>` may leak a bit too much about our implementation and can make future refactorings harder to do without breaking changes.
I don't oppose this addition, but consider it carefully. :pray:

Reply:
Done re: comment.

I'd consider moving to a `HashMap<String, Vec<HashSet<String>>>`, but we'll always be exposing a map of metric to set of sets, so whatever the most generic way to provide that is probably the way to go (or provide less information, like a `HashMap<String, usize>`, but I think I'd prefer giving users the actual tag sets).

Reply:
We could also create a very messed up return type with somethings like `impl Index<str, impl IntoIterator<Item=&str>>`

Reply:
Since actions are only run once, we don't really care about exposing e.g. some data structure we might have to do work to construct in the future (e.g. is we are storing the sets in a `BTreeSet`, handing the user a more generic `HashSet` should still be fine, since we can construct the former given the latter), an this can be more or less arbitrarily expensive.

The thing I think really needs considering here is how the information itself ties us down (rather than the shape the information is given in) - e.g. what if we figure out a way to not need to store the tags in the future? We would still be forced to if we give the users the tag sets.

Given how small the performance impact we've seen with the tag tracking as is, and the fact the number of tag sets is bounded by `threshold`, I'm ok with exposing the tag sets as part of the public interface, and I've made it so that the custom actions get a `&HashMap<String, Vec<HashSet<String>>>`, since that's a pretty close match to what we use internally, but is generic and common enough to be easy for users to reason about. This necessitates a slightly ugly iter chain, but I think it makes the interface generally more friendly.